File: converted_source/upp_rea.pas.py
# unit upp_rea

from conversion_common import *
import udebug
import udomain
import usupport
import uunits
import umath
import updcom
import umain
import usliders
import uppanel
import uparams
import uplant
import delphi_compatability

# const
kItemSlider = 1
kItemValueText = 2
kItemUnitText = 3

class PdRealParameterPanel(PdParameterPanel):
    def __init__(self):
        self.currentValue = 0.0
        self.currentValueWhileSliding = 0.0
        self.minValue = 0.0
        self.maxValue = 0.0
        self.slider = KfSlider()
        self.currentUnitIndex = 0
        self.valueTextRect = TRect()
        self.unitTextRect = TRect()
        self.minTextRect = TRect()
        self.maxTextRect = TRect()
        self.drawWithoutBounds = false
        self.arrayIndex = 0
    
    def initialize(self):
        #assuming slider will be deleted automatically by owner - self - otherwise would have destructor
        self.slider = usliders.KfSlider().create(self)
        self.slider.Parent = self
        self.slider.FOnMouseDown = self.sliderMouseDown
        self.slider.FOnMouseMove = self.sliderMouseMove
        self.slider.FOnMouseUp = self.sliderMouseUp
        self.slider.FOnKeyDown = self.sliderKeyDown
        # make it a percentage move 
        self.slider.maxValue = 100
        self.slider.minValue = 0
        self.slider.readOnly = self.param.readOnly
        self.slider.useDefaultSizeAndDraggerSize()
        self.arrayIndex = uparams.kNotArray
        if udomain.domain.options.useMetricUnits:
            self.currentUnitIndex = self.param.unitMetric
        else:
            self.currentUnitIndex = self.param.unitEnglish
        self.minValue = self.param.lowerBound()
        self.maxValue = self.param.upperBound()
    
    def adjustValue(self):
        newString = ansistring()
        oldValueInCurrentUnit = 0.0
        newValue = 0.0
        newValueInPlantUnit = 0.0
        oldString = ""
        prompt = ""
        nameString = ""
        unitString = ""
        minString = ""
        maxString = ""
        outOfRange = false
        
        oldValueInCurrentUnit = self.toCurrentUnit(self.currentValue)
        newString = usupport.digitValueString(oldValueInCurrentUnit)
        oldString = newString
        nameString = UNRESOLVED.copy(self.Caption, 1, 30)
        if len(nameString) == 30:
            nameString = nameString + "..."
        unitString = uunits.UnitStringForEnum(self.param.unitSet, self.currentUnitIndex)
        minString = usupport.digitValueString(self.toCurrentUnit(self.minValue))
        maxString = usupport.digitValueString(self.toCurrentUnit(self.maxValue))
        prompt = "Type a new value."
        if InputQuery("Enter value", prompt, newString):
            if (newString != oldString) and (usupport.boundForString(newString, self.param.fieldType, newValue)):
                newValueInPlantUnit = self.toPlantUnit(newValue)
                outOfRange = self.checkValueAgainstBounds(newValueInPlantUnit)
                if not outOfRange:
                    self.currentValue = newValueInPlantUnit
                    umain.MainForm.doCommand(updcom.PdChangeRealValueCommand().createCommandWithListOfPlants(umain.MainForm.selectedPlants, self.currentValue, self.param.fieldNumber, self.arrayIndex, self.param.regrow))
                    self.updateDisplay()
                    self.Invalidate()
    
    def toPlantUnit(self, value):
        result = 0.0
        result = uunits.Convert(self.param.unitSet, self.currentUnitIndex, self.param.unitModel, value)
        return result
    
    def toCurrentUnit(self, value):
        result = 0.0
        result = uunits.Convert(self.param.unitSet, self.param.unitModel, self.currentUnitIndex, value)
        return result
    
    def checkValueAgainstBounds(self, aValue):
        result = false
        result = (aValue < self.minValue) or (aValue > self.maxValue)
        return result
    
    def selectNextUnitInSet(self, shift, ctrl):
        if shift:
            self.currentUnitIndex = uunits.GetPreviousUnitEnumInUnitSet(self.param.unitSet, self.currentUnitIndex)
        else:
            self.currentUnitIndex = uunits.GetNextUnitEnumInUnitSet(self.param.unitSet, self.currentUnitIndex)
        self.updateCurrentValue(-1)
        self.updateDisplay()
        self.Invalidate()
    
    def doMouseUp(self, Sender, Button, Shift, X, Y):
        thePoint = TPoint()
        
        # must always call this first because it sets the focus 
        PdParameterPanel.doMouseUp(self, Sender, Button, Shift, X, Y)
        thePoint = Point(X, Y)
        if delphi_compatability.PtInRect(self.valueTextRect, thePoint):
            if self.slider.Enabled and not self.slider.readOnly:
                self.selectedItemIndex = kItemValueText
                self.Invalidate()
                self.adjustValue()
        elif delphi_compatability.PtInRect(self.unitTextRect, thePoint):
            self.selectedItemIndex = kItemUnitText
            self.Invalidate()
            self.selectNextUnitInSet(delphi_compatability.TShiftStateEnum.ssShift in Shift, delphi_compatability.TShiftStateEnum.ssCtrl in Shift)
    
    def doKeyDown(self, sender, key, shift):
        if self.slider.Enabled and not self.slider.readOnly:
            if key == delphi_compatability.VK_HOME:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_END:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_DOWN:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_LEFT:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_UP:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_RIGHT:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_NEXT:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_PRIOR:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
        PdParameterPanel.doKeyDown(self, sender, key, shift)
        if (key == delphi_compatability.VK_RETURN):
            if self.selectedItemIndex == kItemSlider:
                if self.slider.Enabled and not self.slider.readOnly:
                    key = self.slider.doKeyDown(sender, key, shift)
            elif self.selectedItemIndex == kItemValueText:
                if self.slider.Enabled and not self.slider.readOnly:
                    self.adjustValue()
            elif self.selectedItemIndex == kItemUnitText:
                self.selectNextUnitInSet(delphi_compatability.TShiftStateEnum.ssShift in shift, delphi_compatability.TShiftStateEnum.ssCtrl in shift)
        return key
    
    def sliderPositionFromValue(self, value):
        result = 0
        if self.maxValue - self.minValue == 0.0:
            result = 0
        elif value <= self.minValue:
            result = 0
        elif value >= self.maxValue:
            result = 100
        else:
            result = intround(100.0 * (value - self.minValue) / (self.maxValue - self.minValue))
        return result
    
    def valueFromSliderPosition(self):
        result = 0.0
        sliderPosition = 0.0
        
        sliderPosition = self.slider.currentValue
        result = self.minValue + sliderPosition / 100.0 * (self.maxValue - self.minValue)
        return result
    
    def updateEnabling(self):
        PdParameterPanel.updateEnabling(self)
        self.slider.Enabled = self.plant != None
        self.slider.readOnly = self.readOnly
    
    def updatePlantValues(self):
        oldValue = 0.0
        
        if self.plant == None:
            return
        if self.param.collapsed:
            return
        oldValue = self.currentValue
        self.updateCurrentValue(-1)
        if oldValue != self.currentValue:
            self.updateDisplay()
            self.Invalidate()
    
    def updateCurrentValue(self, aFieldIndex):
        if (self.plant != None) and (self.param.fieldType == uparams.kFieldFloat):
            self.currentValue = self.plant.transferField(umath.kGetField, self.currentValue, self.param.fieldNumber, uparams.kFieldFloat, self.arrayIndex, false, None)
            if self.currentValue < self.minValue:
                self.currentValue = self.minValue
            if self.currentValue > self.maxValue:
                self.currentValue = self.maxValue
        else:
            self.currentValue = 0.0
    
    def updateDisplay(self):
        if (self.plant != None) and (self.param.fieldType == uparams.kFieldFloat):
            self.slider.currentValue = self.sliderPositionFromValue(self.currentValue)
        else:
            self.slider.currentValue = 0
            self.slider.Enabled = false
    
    def sliderMouseDown(self, Sender):
        if self.plant == None:
            return
        self.selectedItemIndex = kItemSlider
        self.slider.hasUnofficialFocus = true
        self.currentValueWhileSliding = self.valueFromSliderPosition()
        self.Invalidate()
        if not self.Focused():
            self.SetFocus()
    
    def sliderMouseMove(self, Sender):
        if self.plant == None:
            return
        self.currentValueWhileSliding = self.valueFromSliderPosition()
        self.Invalidate()
    
    def sliderMouseUp(self, Sender):
        if self.plant == None:
            return
        self.currentValue = self.valueFromSliderPosition()
        umain.MainForm.doCommand(updcom.PdChangeRealValueCommand().createCommandWithListOfPlants(umain.MainForm.selectedPlants, self.currentValue, self.param.fieldNumber, self.arrayIndex, self.param.regrow))
        self.Invalidate()
    
    def sliderKeyDown(self, Sender):
        self.sliderMouseUp(Sender)
    
    def maxWidth(self):
        result = 0
        result = umath.intMax(uppanel.kLeftRightGap * 2 + self.labelWidth, self.minScaleWidthWithBounds(self.param.unitSet, self.param.unitModel, self.minValue, self.maxValue))
        return result
    
    def minWidth(self, requestedWidth):
        result = 0
        widthForLongestWord = 0
        widthWithBounds = 0
        widthWithoutBounds = 0
        
        # 1. test if width given is less than labelWidth.
        #       if not, move on
        #       if so, test if width given is less than width of longest full word in labelWidth.
        #         if not, move on
        #         if so, give back width of longest full word in labelWidth as minWidth.
        #    2. test if width given is less than minScaleWidthWithBounds.
        #       if not, move on (set flag to draw bounds)
        #       if so, test if width given is less than minScaleWidthWithoutBounds.
        #         if not, move on (set flag to not draw bounds)
        #         if so, give back minScaleWidthWithoutBounds as minWidth.
        #  
        widthForLongestWord = uppanel.kLeftRightGap * 2 + self.longestLabelWordWidth
        widthWithBounds = self.minScaleWidthWithBounds(self.param.unitSet, self.param.unitModel, self.minValue, self.maxValue)
        widthWithoutBounds = self.minScaleWidthWithoutBounds(self.param.unitSet, self.param.unitModel)
        result = -1
        if requestedWidth < widthForLongestWord:
            result = widthForLongestWord
        if requestedWidth < widthWithBounds:
            self.drawWithoutBounds = true
            if requestedWidth < widthWithoutBounds:
                result = umath.intMax(result, widthWithoutBounds)
        else:
            self.drawWithoutBounds = false
        return result
    
    def uncollapsedHeight(self):
        result = 0
        result = self.collapsedHeight() + self.textHeight + uppanel.kTopBottomGap
        return result
    
    def resizeElements(self):
        self.paint()
        self.slider.Left = self.minTextRect.Right + uppanel.kSliderSpaceGap
        self.slider.Top = self.collapsedHeight() + self.textHeight / 2 - self.slider.Height / 2
        self.slider.Width = umath.intMax(0, self.maxTextRect.Left - self.minTextRect.Right - uppanel.kSliderSpaceGap * 2)
    
    def maxSelectedItemIndex(self):
        result = 0
        if (not self.param.collapsed) and (self.plant != None):
            result = kItemUnitText
        else:
            result = uppanel.kItemLabel
        return result
    
    def nextSelectedItemIndex(self, goForward):
        result = 0
        result = uppanel.kItemNone
        if self.selectedItemIndex == uppanel.kItemNone:
            result = self.selectedItemIndex
        elif self.selectedItemIndex == uppanel.kItemLabel:
            if goForward:
                result = kItemValueText
            else:
                result = uppanel.kItemNone
        elif self.selectedItemIndex == kItemSlider:
            if goForward:
                result = uppanel.kItemNone
            else:
                result = kItemUnitText
        elif self.selectedItemIndex == kItemValueText:
            if goForward:
                result = kItemUnitText
            else:
                result = uppanel.kItemLabel
        elif self.selectedItemIndex == kItemUnitText:
            if goForward:
                result = kItemSlider
            else:
                result = kItemValueText
        return result
    
    def paint(self):
        Rect = TRect()
        minString = ""
        valueString = ""
        maxString = ""
        unitString = ""
        
        if delphi_compatability.Application.terminated:
            return
        if (self.selectedItemIndex == kItemSlider):
            # ask slider to update itself based on whether it is selected 
            self.slider.hasUnofficialFocus = true
        else:
            self.slider.hasUnofficialFocus = false
        PdParameterPanel.paint(self)
        self.slider.Visible = not self.param.collapsed
        if (self.param.collapsed):
            self.drawExtraValueCaptionWithString(usupport.digitValueString(self.toCurrentUnit(self.currentValue)), uunits.UnitStringForEnum(self.param.unitSet, self.currentUnitIndex))
            return
        Rect = self.GetClientRect()
        minString = usupport.digitValueString(self.toCurrentUnit(self.minValue))
        maxString = usupport.digitValueString(self.toCurrentUnit(self.maxValue))
        if self.slider.dragging:
            valueString = usupport.digitValueString(self.toCurrentUnit(self.currentValueWhileSliding))
        else:
            valueString = usupport.digitValueString(self.toCurrentUnit(self.currentValue))
        unitString = uunits.UnitStringForEnum(self.param.unitSet, self.currentUnitIndex)
        self.valueTextRect.Left = uppanel.kLeftRightGap
        self.valueTextRect.Right = self.valueTextRect.Left + self.Canvas.TextWidth(valueString)
        self.valueTextRect.Top = self.collapsedHeight() + 1
        self.valueTextRect.Bottom = self.valueTextRect.Top + self.textHeight
        self.unitTextRect = self.valueTextRect
        self.unitTextRect.Left = self.valueTextRect.Right + 5
        self.unitTextRect.Right = self.unitTextRect.Left + self.Canvas.TextWidth(unitString)
        self.minTextRect = self.valueTextRect
        self.minTextRect.Left = self.unitTextRect.Right + 5
        self.minTextRect.Right = self.minTextRect.Left + self.Canvas.TextWidth(minString)
        self.maxTextRect = self.valueTextRect
        self.maxTextRect.Right = Rect.right - uppanel.kLeftRightGap
        self.maxTextRect.Left = self.maxTextRect.Right - self.Canvas.TextWidth(maxString)
        self.drawText(minString, self.minTextRect, false, false, true)
        self.drawText(valueString, self.valueTextRect, not self.readOnly, (self.selectedItemIndex == kItemValueText), false)
        self.drawText(unitString, self.unitTextRect, true, (self.selectedItemIndex == kItemUnitText), false)
        self.drawText(maxString, self.maxTextRect, false, false, true)
    

File: converted_source/upp_rea.pas.java
// unit upp_rea

from conversion_common import *;
import udebug;
import usliders;
import uppanel;
import delphi_compatability;

// const
kItemSlider = 1;
kItemValueText = 2;
kItemUnitText = 3;



class PdRealParameterPanel extends PdParameterPanel {
    public float currentValue;
    public float currentValueWhileSliding;
    public float minValue;
    public float maxValue;
    public KfSlider slider;
    public int currentUnitIndex;
    public TRect valueTextRect;
    public TRect unitTextRect;
    public TRect minTextRect;
    public TRect maxTextRect;
    public boolean drawWithoutBounds;
    public short arrayIndex;
    
    public void initialize() {
        //assuming slider will be deleted automatically by owner - self - otherwise would have destructor
        this.slider = usliders.KfSlider().create(this);
        this.slider.Parent = this;
        this.slider.FOnMouseDown = this.sliderMouseDown;
        this.slider.FOnMouseMove = this.sliderMouseMove;
        this.slider.FOnMouseUp = this.sliderMouseUp;
        this.slider.FOnKeyDown = this.sliderKeyDown;
        // make it a percentage move 
        this.slider.maxValue = 100;
        this.slider.minValue = 0;
        this.slider.readOnly = this.param.readOnly;
        this.slider.useDefaultSizeAndDraggerSize();
        this.arrayIndex = UNRESOLVED.kNotArray;
        if (UNRESOLVED.domain.options.useMetricUnits) {
            this.currentUnitIndex = this.param.unitMetric;
        } else {
            this.currentUnitIndex = this.param.unitEnglish;
        }
        this.minValue = this.param.lowerBound;
        this.maxValue = this.param.upperBound;
    }
    
    public void adjustValue() {
        ansistring newString = new ansistring();
        float oldValueInCurrentUnit = 0.0;
        float newValue = 0.0;
        float newValueInPlantUnit = 0.0;
        String oldString = "";
        String prompt = "";
        String nameString = "";
        String unitString = "";
        String minString = "";
        String maxString = "";
        boolean outOfRange = false;
        
        oldValueInCurrentUnit = this.toCurrentUnit(this.currentValue);
        newString = UNRESOLVED.digitValueString(oldValueInCurrentUnit);
        oldString = newString;
        nameString = UNRESOLVED.copy(this.Caption, 1, 30);
        if (len(nameString) == 30) {
            nameString = nameString + "...";
        }
        unitString = UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.currentUnitIndex);
        minString = UNRESOLVED.digitValueString(this.toCurrentUnit(this.minValue));
        maxString = UNRESOLVED.digitValueString(this.toCurrentUnit(this.maxValue));
        prompt = "Type a new value.";
        if (InputQuery("Enter value", prompt, newString)) {
            if ((newString != oldString) && (UNRESOLVED.boundForString(newString, this.param.fieldType, newValue))) {
                newValueInPlantUnit = this.toPlantUnit(newValue);
                outOfRange = this.checkValueAgainstBounds(newValueInPlantUnit);
                if (!outOfRange) {
                    this.currentValue = newValueInPlantUnit;
                    UNRESOLVED.MainForm.doCommand(UNRESOLVED.PdChangeRealValueCommand.createCommandWithListOfPlants(UNRESOLVED.MainForm.selectedPlants, this.currentValue, this.param.fieldNumber, this.arrayIndex, this.param.regrow));
                    this.updateDisplay();
                    this.Invalidate();
                }
            }
        }
    }
    
    public float toPlantUnit(float value) {
        result = 0.0;
        result = UNRESOLVED.Convert(this.param.unitSet, this.currentUnitIndex, this.param.unitModel, value);
        return result;
    }
    
    public float toCurrentUnit(float value) {
        result = 0.0;
        result = UNRESOLVED.Convert(this.param.unitSet, this.param.unitModel, this.currentUnitIndex, value);
        return result;
    }
    
    public boolean checkValueAgainstBounds(float aValue) {
        result = false;
        result = (aValue < this.minValue) || (aValue > this.maxValue);
        return result;
    }
    
    public void selectNextUnitInSet(boolean shift, boolean ctrl) {
        if (shift) {
            this.currentUnitIndex = UNRESOLVED.GetPreviousUnitEnumInUnitSet(this.param.unitSet, this.currentUnitIndex);
        } else {
            this.currentUnitIndex = UNRESOLVED.GetNextUnitEnumInUnitSet(this.param.unitSet, this.currentUnitIndex);
        }
        this.updateCurrentValue(-1);
        this.updateDisplay();
        this.Invalidate();
    }
    
    public void doMouseUp(TObject Sender, TMouseButton Button, TShiftState Shift, int X, int Y) {
        TPoint thePoint = new TPoint();
        
        // must always call this first because it sets the focus 
        super.doMouseUp(Sender, Button, Shift, X, Y);
        thePoint = Point(X, Y);
        if (delphi_compatability.PtInRect(this.valueTextRect, thePoint)) {
            if (this.slider.Enabled && !this.slider.readOnly) {
                this.selectedItemIndex = kItemValueText;
                this.Invalidate();
                this.adjustValue();
            }
        } else if (delphi_compatability.PtInRect(this.unitTextRect, thePoint)) {
            this.selectedItemIndex = kItemUnitText;
            this.Invalidate();
            this.selectNextUnitInSet(delphi_compatability.TShiftStateEnum.ssShift in Shift, delphi_compatability.TShiftStateEnum.ssCtrl in Shift);
        }
    }
    
    public void doKeyDown(TObject sender, byte key, TShiftState shift) {
        if (this.slider.Enabled && !this.slider.readOnly) {
            switch (key) {
                case delphi_compatability.VK_HOME:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_END:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_DOWN:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_LEFT:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_UP:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_RIGHT:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_NEXT:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_PRIOR:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
        }
        super.doKeyDown(sender, key, shift);
        if ((key == delphi_compatability.VK_RETURN)) {
            switch (this.selectedItemIndex) {
                case kItemSlider:
                    if (this.slider.Enabled && !this.slider.readOnly) {
                        key = this.slider.doKeyDown(sender, key, shift);
                    }
                    break;
                case kItemValueText:
                    if (this.slider.Enabled && !this.slider.readOnly) {
                        this.adjustValue();
                    }
                    break;
                case kItemUnitText:
                    this.selectNextUnitInSet(delphi_compatability.TShiftStateEnum.ssShift in shift, delphi_compatability.TShiftStateEnum.ssCtrl in shift);
                    break;
        }
        return key;
    }
    
    public int sliderPositionFromValue(float value) {
        result = 0;
        if (this.maxValue - this.minValue == 0.0) {
            result = 0;
        } else if (value <= this.minValue) {
            result = 0;
        } else if (value >= this.maxValue) {
            result = 100;
        } else {
            result = intround(100.0 * (value - this.minValue) / (this.maxValue - this.minValue));
        }
        return result;
    }
    
    public float valueFromSliderPosition() {
        result = 0.0;
        float sliderPosition = 0.0;
        
        sliderPosition = this.slider.currentValue;
        result = this.minValue + sliderPosition / 100.0 * (this.maxValue - this.minValue);
        return result;
    }
    
    public void updateEnabling() {
        super.updateEnabling();
        this.slider.Enabled = this.plant != null;
        this.slider.readOnly = this.readOnly;
    }
    
    public void updatePlantValues() {
        float oldValue = 0.0;
        
        if (this.plant == null) {
            return;
        }
        if (this.param.collapsed) {
            return;
        }
        oldValue = this.currentValue;
        this.updateCurrentValue(-1);
        if (oldValue != this.currentValue) {
            this.updateDisplay();
            this.Invalidate();
        }
    }
    
    public void updateCurrentValue(int aFieldIndex) {
        if ((this.plant != null) && (this.param.fieldType == UNRESOLVED.kFieldFloat)) {
            this.plant.transferField(UNRESOLVED.kGetField, this.currentValue, this.param.fieldNumber, UNRESOLVED.kFieldFloat, this.arrayIndex, false, null);
            if (this.currentValue < this.minValue) {
                this.currentValue = this.minValue;
            }
            if (this.currentValue > this.maxValue) {
                this.currentValue = this.maxValue;
            }
        } else {
            this.currentValue = 0.0;
        }
    }
    
    public void updateDisplay() {
        if ((this.plant != null) && (this.param.fieldType == UNRESOLVED.kFieldFloat)) {
            this.slider.currentValue = this.sliderPositionFromValue(this.currentValue);
        } else {
            this.slider.currentValue = 0;
            this.slider.Enabled = false;
        }
    }
    
    public void sliderMouseDown(TObject Sender) {
        if (this.plant == null) {
            return;
        }
        this.selectedItemIndex = kItemSlider;
        this.slider.hasUnofficialFocus = true;
        this.currentValueWhileSliding = this.valueFromSliderPosition();
        this.Invalidate();
        if (!this.Focused()) {
            this.SetFocus();
        }
    }
    
    public void sliderMouseMove(TObject Sender) {
        if (this.plant == null) {
            return;
        }
        this.currentValueWhileSliding = this.valueFromSliderPosition();
        this.Invalidate();
    }
    
    public void sliderMouseUp(TObject Sender) {
        if (this.plant == null) {
            return;
        }
        this.currentValue = this.valueFromSliderPosition();
        UNRESOLVED.MainForm.doCommand(UNRESOLVED.PdChangeRealValueCommand.createCommandWithListOfPlants(UNRESOLVED.MainForm.selectedPlants, this.currentValue, this.param.fieldNumber, this.arrayIndex, this.param.regrow));
        this.Invalidate();
    }
    
    public void sliderKeyDown(TObject Sender) {
        this.sliderMouseUp(Sender);
    }
    
    public int maxWidth() {
        result = 0;
        result = UNRESOLVED.intMax(uppanel.kLeftRightGap * 2 + this.labelWidth, this.minScaleWidthWithBounds(this.param.unitSet, this.param.unitModel, this.minValue, this.maxValue));
        return result;
    }
    
    public int minWidth(int requestedWidth) {
        result = 0;
        int widthForLongestWord = 0;
        int widthWithBounds = 0;
        int widthWithoutBounds = 0;
        
        // 1. test if width given is less than labelWidth.
        //       if not, move on
        //       if so, test if width given is less than width of longest full word in labelWidth.
        //         if not, move on
        //         if so, give back width of longest full word in labelWidth as minWidth.
        //    2. test if width given is less than minScaleWidthWithBounds.
        //       if not, move on (set flag to draw bounds)
        //       if so, test if width given is less than minScaleWidthWithoutBounds.
        //         if not, move on (set flag to not draw bounds)
        //         if so, give back minScaleWidthWithoutBounds as minWidth.
        //  
        widthForLongestWord = uppanel.kLeftRightGap * 2 + this.longestLabelWordWidth;
        widthWithBounds = this.minScaleWidthWithBounds(this.param.unitSet, this.param.unitModel, this.minValue, this.maxValue);
        widthWithoutBounds = this.minScaleWidthWithoutBounds(this.param.unitSet, this.param.unitModel);
        result = -1;
        if (requestedWidth < widthForLongestWord) {
            result = widthForLongestWord;
        }
        if (requestedWidth < widthWithBounds) {
            this.drawWithoutBounds = true;
            if (requestedWidth < widthWithoutBounds) {
                result = UNRESOLVED.intMax(result, widthWithoutBounds);
            }
        } else {
            this.drawWithoutBounds = false;
        }
        return result;
    }
    
    public int uncollapsedHeight() {
        result = 0;
        result = this.collapsedHeight() + this.textHeight + uppanel.kTopBottomGap;
        return result;
    }
    
    public void resizeElements() {
        this.paint();
        this.slider.Left = this.minTextRect.Right + uppanel.kSliderSpaceGap;
        this.slider.Top = this.collapsedHeight() + this.textHeight / 2 - this.slider.Height / 2;
        this.slider.Width = UNRESOLVED.intMax(0, this.maxTextRect.Left - this.minTextRect.Right - uppanel.kSliderSpaceGap * 2);
    }
    
    public int maxSelectedItemIndex() {
        result = 0;
        if ((!this.param.collapsed) && (this.plant != null)) {
            result = kItemUnitText;
        } else {
            result = uppanel.kItemLabel;
        }
        return result;
    }
    
    public int nextSelectedItemIndex(boolean goForward) {
        result = 0;
        result = uppanel.kItemNone;
        switch (this.selectedItemIndex) {
            case uppanel.kItemNone:
                result = this.selectedItemIndex;
                break;
            case uppanel.kItemLabel:
                if (goForward) {
                    result = kItemValueText;
                } else {
                    result = uppanel.kItemNone;
                }
                break;
            case kItemSlider:
                if (goForward) {
                    result = uppanel.kItemNone;
                } else {
                    result = kItemUnitText;
                }
                break;
            case kItemValueText:
                if (goForward) {
                    result = kItemUnitText;
                } else {
                    result = uppanel.kItemLabel;
                }
                break;
            case kItemUnitText:
                if (goForward) {
                    result = kItemSlider;
                } else {
                    result = kItemValueText;
                }
                break;
        return result;
    }
    
    public void paint() {
        TRect Rect = new TRect();
        String minString = "";
        String valueString = "";
        String maxString = "";
        String unitString = "";
        
        if (delphi_compatability.Application.terminated) {
            return;
        }
        if ((this.selectedItemIndex == kItemSlider)) {
            // ask slider to update itself based on whether it is selected 
            this.slider.hasUnofficialFocus = true;
        } else {
            this.slider.hasUnofficialFocus = false;
        }
        super.paint();
        this.slider.Visible = !this.param.collapsed;
        if ((this.param.collapsed)) {
            this.drawExtraValueCaptionWithString(UNRESOLVED.digitValueString(this.toCurrentUnit(this.currentValue)), UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.currentUnitIndex));
            return;
        }
        Rect = this.GetClientRect();
        minString = UNRESOLVED.digitValueString(this.toCurrentUnit(this.minValue));
        maxString = UNRESOLVED.digitValueString(this.toCurrentUnit(this.maxValue));
        if (this.slider.dragging) {
            valueString = UNRESOLVED.digitValueString(this.toCurrentUnit(this.currentValueWhileSliding));
        } else {
            valueString = UNRESOLVED.digitValueString(this.toCurrentUnit(this.currentValue));
        }
        unitString = UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.currentUnitIndex);
        this.valueTextRect.Left = uppanel.kLeftRightGap;
        this.valueTextRect.Right = this.valueTextRect.Left + this.Canvas.TextWidth(valueString);
        this.valueTextRect.Top = this.collapsedHeight() + 1;
        this.valueTextRect.Bottom = this.valueTextRect.Top + this.textHeight;
        this.unitTextRect = this.valueTextRect;
        this.unitTextRect.Left = this.valueTextRect.Right + 5;
        this.unitTextRect.Right = this.unitTextRect.Left + this.Canvas.TextWidth(unitString);
        this.minTextRect = this.valueTextRect;
        this.minTextRect.Left = this.unitTextRect.Right + 5;
        this.minTextRect.Right = this.minTextRect.Left + this.Canvas.TextWidth(minString);
        this.maxTextRect = this.valueTextRect;
        this.maxTextRect.Right = Rect.right - uppanel.kLeftRightGap;
        this.maxTextRect.Left = this.maxTextRect.Right - this.Canvas.TextWidth(maxString);
        this.drawText(minString, this.minTextRect, false, false, true);
        this.drawText(valueString, this.valueTextRect, !this.readOnly, (this.selectedItemIndex == kItemValueText), false);
        this.drawText(unitString, this.unitTextRect, true, (this.selectedItemIndex == kItemUnitText), false);
        this.drawText(maxString, this.maxTextRect, false, false, true);
    }
    
}
